solidity/ecdsa/contracts/WalletRegistry.sol

```diff
@@ -0,0 +1,327 @@
+    );
+
+    constructor(
+        SortitionPool _sortitionPool,
+        IERC20 _tToken,
+        IWalletStaking _staking,
+        DKGValidator _dkgValidator,
+        address _walletOwner
+    ) {
+        sortitionPool = _sortitionPool;
+        tToken = _tToken;
+        staking = _staking;
+        walletOwner = _walletOwner;
+
+        // TODO: Implement governance for the parameters
+        // TODO: revisit all initial values
+
+        maliciousDkgResultSlashingAmount = 50000e18;
+        maliciousDkgResultNotificationRewardMultiplier = 100;
+
+        dkg.init(_sortitionPool, _dkgValidator);
+        dkg.setResultChallengePeriodLength(11520); // ~48h assuming 15s block time
+        dkg.setResultSubmissionEligibilityDelay(20);
+    }
+
+    // TODO: Update to governable params
+    function updateDkgParams(
+        uint256 newResultChallengePeriodLength,
+        uint256 newResultSubmissionEligibilityDelay
+    ) external {
+        dkg.setResultChallengePeriodLength(newResultChallengePeriodLength);
+        dkg.setResultSubmissionEligibilityDelay(
+            newResultSubmissionEligibilityDelay
+        );
+    }
+
+    /// @notice Registers the caller in the sortition pool.
+    function registerOperator() external {
+        address operator = msg.sender;
+
+        require(
+            !sortitionPool.isOperatorInPool(operator),
+            "Operator is already registered"
+        );
+
+        sortitionPool.insertOperator(
+            operator,
+            staking.eligibleStake(operator, address(this))
+        );
+    }
+
+    /// @notice Updates the sortition pool status of the caller.
+    function updateOperatorStatus() external {
+        sortitionPool.updateOperatorStatus(
+            msg.sender,
+            staking.eligibleStake(msg.sender, address(this))
+        );
+    }
+
+    /// @notice Requests a new wallet creation.
+    /// @dev Can be called only by the owner of wallets.
+    ///      It starts a new DKG process.
+    function requestNewWallet() external {
+        require(msg.sender == walletOwner, "Caller is not the Wallet Owner");
+
+        dkg.lockState();
+        dkg.start(
+            uint256(keccak256(abi.encodePacked(relayEntry, block.number)))
+        );
+    }
+
+    /// @notice Submits result of DKG protocol. It is on-chain part of phase 14 of
+    ///         the protocol. The DKG result consists of result submitting member
+    ///         index, calculated group public key, bytes array of misbehaved
+    ///         members, concatenation of signatures from group members,
+    ///         indices of members corresponding to each signature and
+    ///         the list of group members.
+    ///         When the result is verified successfully it gets registered and
+    ///         waits for an approval. A result can be challenged to verify the
+    ///         members list corresponds to the expected set of members determined
+    ///         by the sortition pool.
```

Comment:
> I think we could improve the docs. The result should be challenged when something is wrong with it and now it could be more than just members determined by the sortition pool.

f194f636844419ce7850e4a61d57284c0d72e716

Reply:
> Also, we could expose a function in WalletRegistry allowing for a simple true/false verification of the result using the current entry value (with a try-catch). This would simplify the code in the client. It is not in the RandomBeacon given the contract size constraints but we should be able to add it here. I'd also love to add it to RandomBeacon once we ensure we have enough bytecode left there after integrating with T staking contract.

We could achieve the same result in the client by simulating a non-mutating call to the `challengeDkgResult` function, IMHO there is no need to expose it with another try/catch function. Our code Go code generator  exposes such call functions, see example here: https://github.com/keep-network/keep-ecdsa/blob/2cc8e0b89043e3c12eedb8c3ce46734c2f3fd2d6/pkg/chain/gen/celo/contract/BondedECDSAKeep.go#L200-L203

Reply:
The call from the client can fail because of revert in the validator and because of some unrelated/infra-level problem. If we do not expose this function in a smart contract, we will have to implement various paths based on errors returned and we'll never have 100% correctness.

Reply:
If we expose the function in a smart contract it can also fail because of some unrelated/infra-level problem, can't it? Sorry, I may be missing a specific scenario you have in mind.

Reply:
DKG result submitted may contain corrupted data leading `DKGValidator.validate` to revert, that's why we have a try-catch in `DKG.challengeResult` and consider a revert from `DKGValidator.validate` as a malicious submission.

If we have a function in `WalletRegistry.isResultValid(DKGResult): bool` with a try-catch inside, we can simplify the off-chain client code in the following way:
- Group member before submitting DKG result to the chain calls `WalletRegistry.isResultValid(DKGResult)`. If the function returned true, it is submitting the result. If the function failed with an error, the call is retried given it must be an infra-level problem. If the function returned false, it is not submitting the result (something terrible happened during off-chain protocol).
- Anyone: upon receiving an event about DKG result submitted, call `WalletRegistry.isResultValid(DKGResult)`. If the function returned true, do nothing. If the function returned false, submit challenge. If the function failed, retry given it must be an infra-level problem.

Reply:
Let's remember about this function ☝🏻

Reply:
22695d96b23f36b7989f91494c2b2f3867b6433b